### PR TITLE
fix: commands shouldn't be in ASC order

### PR DESCRIPTION
**Pull request by PaulDremanovich**
https://trello.com/c/oMvsGyEs/527-commands-shouldnt-be-in-asc-order

- The last command entered must be the first in the list.
- Load command history on start

[Screencast from 26.03.2024 15:44:27.webm](https://github.com/Adamant-im/adamant-im/assets/160286261/48a1c6e3-948c-499d-aa4e-5d5b85569616)

### CODE REVIEW

**Comment by PaulDremanovich** [comment]
[Commands shouldn't be in ASC order](https://trello.com/c/oMvsGyEs/527-commands-shouldnt-be-in-asc-order)

File: src/store/modules/chat/index.js [hunk 1]
```diff
@@ -558,6 +558,20 @@ const actions = {
       .then(({ messages, lastOffset }) => {
         dispatch('unshiftMessages', messages)
 
+        if (messages && offset === 0) {
+          dispatch(
+            'botCommands/initBotCommands',
+            {
+              partnerId: contactId,
+              commands: messages
+                .filter((item) => item.message.startsWith('/'))
```
Comment (bludnic) [inline src/store/modules/chat/index.js:567]:
```suggestion
                .filter((item) => item.message?.startsWith('/'))
```

Message may be undefined

File: src/store/modules/chat/index.js [hunk 2]
```diff
@@ -564,7 +564,7 @@ const actions = {
             {
               partnerId: contactId,
               commands: messages
-                .filter((item) => item.message.startsWith('/'))
+                .filter((item) => item.message?.startsWith('/'))
```
Comment (bludnic) [inline src/store/modules/chat/index.js:567]: Verify using `typeof string`, because in case of crypto transfer, the `message` is an object type

image

File: src/store/modules/bot-commands/bot-commands-mutations.ts [hunk 3]
```diff
@@ -1,14 +1,27 @@
...
+      }
       botCommands.push(commandValue)
     }
+    state.commands[botCommand.partnerId] = botCommands
+  },
+
+  initCommands(state, { partnerId, commands }: { partnerId: string; commands: string[] }): void {
+    const botCommands = state.commands[partnerId]
+    if (!botCommands) {
+      state.commands[partnerId] = uniq(commands)
```
Comment (bludnic) [inline src/store/modules/bot-commands/bot-commands-mutations.ts:24]: Doesn't seem that `uniq()` works as we need:

image

**Expected result:**
...
/orders ETH/USDT
/params
/start mm
/orders
/stop mm

File: src/store/modules/bot-commands/bot-commands-mutations.ts [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+        botCommands.splice(index, 1)
+      }
       botCommands.push(commandValue)
     }
+    state.commands[botCommand.partnerId] = botCommands
+  },
+
+  initCommands(state, { partnerId, commands }: { partnerId: string; commands: string[] }): void {
+    const botCommands = state.commands[partnerId]
+    if (!botCommands) {
```
Comment (bludnic) [inline src/store/modules/bot-commands/bot-commands-mutations.ts:23]: I think you can skip this check and override the commands anyway.

Scenario:

1. You opened a chat in PWA app
2. Initial bot commands were saved to Session Storage
3. From iOS app you send new commands
4. Back to PWA. Refresh the page

**Expected result:** Init commands again

**Actual result:** Initialization is skipped. Using old state from SS

**Review by bludnic — CHANGES_REQUESTED** [review]

File: src/store/modules/draft-message/index.ts [hunk 5]
```diff
@@ -31,11 +31,17 @@ const mutations: MutationTree<DraftState> = {
...
+      delete draft.replyToId
+    }
   },
 
   deleteMessage(state, payload: { partnerId: string; message: string }) {
-    delete state.drafts[payload.partnerId].message
+    const draft = state.drafts[payload.partnerId]
+    if (draft) {
+      delete draft.message
+    }
```
Comment (bludnic) [inline src/store/modules/draft-message/index.ts:44]: Is this related to the current task? Could you describe the bug?

File: src/store/modules/chat/index.js [hunk 6]
```diff
@@ -607,6 +596,28 @@ const actions = {
...
+                partnerId: recipientId,
+                commands
+              },
+              { root: true }
+            )
+          }
+        }
+      }
+    }
+  },
```
Comment (bludnic) [inline src/store/modules/chat/index.js:619]: This logic is more related to `bot-commands` module.
Move the action to `bot-commands/`

File: src/store/modules/chat/index.js [hunk 7]
```diff
@@ -587,7 +574,8 @@ const actions = {
    * Push array of messages and sort by senderId.
    * @param {Message[]} messages Array of messages
    */
-  pushMessages({ commit, rootState }, messages) {
+  pushMessages({ commit, rootState, dispatch }, messages) {
+    dispatch('reInitCommands', messages)
```
Comment (bludnic) [inline src/store/modules/chat/index.js:578]: Are you sure you need the raw messages? Can you use already normalized messages (see 581 line)?

File: src/store/modules/chat/index.js [hunk 8]
```diff
@@ -596,7 +584,8 @@ const actions = {
     })
   },
 
-  unshiftMessages({ commit, rootState }, messages) {
+  unshiftMessages({ commit, rootState, dispatch }, messages) {
+    dispatch('reInitCommands', messages)
```
Comment (bludnic) [inline src/store/modules/chat/index.js:588]: See previous comment

File: src/store/modules/bot-commands/utils/uniqCommand.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { BotCommand } from '@/store/modules/bot-commands/types.ts'
+
+export function uniqCommand(commands: BotCommand[]) {
```
Comment (bludnic) [inline src/store/modules/bot-commands/utils/uniqCommand.ts:3]: Add JSDoc

File: src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { getRealTimestamp } from '@/lib/chat/helpers/utils/getRealTimestamp'
+import { BotCommand } from '@/store/modules/bot-commands/types.ts'
+
+export function extractCommandsFromMessages(
```
Comment (bludnic) [inline src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts:4]: Add JSDoc. Explain what function does.

File: src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import { getRealTimestamp } from '@/lib/chat/helpers/utils/getRealTimestamp'
+import { BotCommand } from '@/store/modules/bot-commands/types.ts'
```
Comment (bludnic) [inline src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts:2]: rm .ts

File: src/store/modules/bot-commands/bot-commands-mutations.ts [hunk 12]
```diff
@@ -1,27 +1,32 @@
 import { MutationTree } from 'vuex'
 import { BotCommand, BotCommandsState } from '@/store/modules/bot-commands/types.ts'
-import { uniq } from 'lodash'
+import { uniqCommand } from '@/store/modules/bot-commands/utils/uniqCommand.ts'
```
Comment (bludnic) [inline src/store/modules/bot-commands/bot-commands-mutations.ts:3]: remove .ts ext

**Review by bludnic — CHANGES_REQUESTED** [review]

**Comment by bludnic** [comment]
Tested the logic. Works well 👍

**Comment by RealGoodProgrammer** [comment]
Agreed with Victor. Logic now works fine

File: src/store/modules/draft-message/index.ts (hunk 5 above)
Comment (PaulDremanovich) [reply src/store/modules/draft-message/index.ts:44]: Bug description:
After some time, if you return to the PWA screen. Then the selected command is not sent when you press Enter. Because when we send a message, we try to remove it from the draft, but the draft is empty at the moment.

File: src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+ * @return {BotCommand[]} Bot commands array
+ * */
+export function extractCommandsFromMessages(
+  recipientId: string,
+  messages: NormalizedChatMessageTransaction[]
+): BotCommand[] {
+  return messages
+    .filter((item) => item.recipientId === recipientId)
+    .filter((item) => typeof item.message === 'string')
+    .filter((item) => item.message?.startsWith('/'))
```
Comment (RealGoodProgrammer) [inline src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts]: Are you sure we need 3 filters here?

**Review by RealGoodProgrammer — APPROVED** [review]

File: src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import { BotCommand } from '@/store/modules/bot-commands/types'
+import { NormalizedChatMessageTransaction } from '@/lib/chat/helpers'
+
+/**
+ * The function retrieves all bot commands from the message array
+ * @param {string} recipientId  Message recipient address
+ * @param {NormalizedChatMessageTransaction[]} messages Array of normalized messages
+ * @return {BotCommand[]} Bot commands array
```
Comment (bludnic) [inline src/store/modules/bot-commands/utils/extractCommandsFromMessages.ts:8]:
```suggestion
 * @param recipientId - Message recipient address
 * @param messages - Array of normalized messages
 * @return Bot commands array
```

You can remove the types from docblock. They are already described by TS.
https://tsdoc.org/

**Review by bludnic — APPROVED** [review]

File: package.json [hunk 15]
```diff
@@ -114,6 +114,7 @@
     "@types/dompurify": "^3.0.5",
     "@types/emoji-mart": "^3.0.14",
     "@types/eslint": "^8.56.6",
+    "@types/lodash": "^4.17.0",
```
Comment (adamant-al) [inline package.json:117]: @bludnic 
Do we really want lodash in our project?

Comment (bludnic) [reply package.json:117]: It was already there. These are just TS types.

Comment (bludnic) [reply package.json:117]: Maybe we will remove lodash in the future. Currently is not possible because we already use it.
